spec/lib/models/i18n_repository_spec.rb

```diff
@@ -0,0 +1,168 @@
+
+  describe '#initialize' do
+    it 'initializes the Crowdin api with correct arguments' do
+      expect(Crowdin::API).to receive(:new).with(project_id: project_id, api_key: api_key)
+      i18n_provider
+    end
+  end
+
+  describe '#create_handle' do
+    it 'returns an instance of Crowdin::API' do
+      expect(i18n_provider.create_handle).to be_an_instance_of(Crowdin::API)
+    end
+  end
+
+  describe '#configuration' do
+    let(:configuration) { {'test' => 'hash' } }
+    it 'uses the LocalesUpdaterConfiguration' do
+      allow(LocalesUpdaterConfiguration).to receive(:configuration).and_return(configuration)
+      expect(i18n_provider.configuration).to eql(configuration)
+    end
+  end
+
+  describe '#upload_english' do
+    let(:crowdin) { instance_double(Crowdin::API) }
+    let(:translation_file) { 'test-file' }
+    let(:path_to_translation) { 'test-path' }
+    let(:title) { 'test-title' }
+
+    it 'creates crowdin directory if missing' do
+      allow(i18n_provider).to receive(:file_exists_in_directory?).and_return(true)
+      allow(i18n_provider).to receive(:update_file)
+
+      expect(i18n_provider).to receive(:add_directory_if_missing)
+      i18n_provider.upload_english(translation_file, path_to_translation, title)
+    end
+
+    context 'with the file already on Crowdin' do
+      before do
+        allow(i18n_provider).to receive(:file_exists_in_directory?).and_return(true)
+      end
+
+      it 'updates the file on Crowdin' do
+        allow(i18n_provider).to receive(:add_directory_if_missing)
+
+        expect(i18n_provider).to receive(:update_file)
+        i18n_provider.upload_english(translation_file, path_to_translation, title)
+      end
+    end
+
+    context 'with the file not already on Crowdin' do
+      before do
+        allow(i18n_provider).to receive(:file_exists_in_directory?).and_return(false)
+      end
+
+      it 'updates the file on Crowdin' do
+        allow(i18n_provider).to receive(:add_directory_if_missing)
+
+        expect(i18n_provider).to receive(:add_file)
+        i18n_provider.upload_english(translation_file, path_to_translation, title)
+      end
+    end
+  end
+
+  describe '#add_directory_if_missing' do
+    context 'with the directory missing on Crowdin' do
+      before do
+        allow(i18n_provider).to receive(:crowdin_directory_exists?).and_return(false)
+        i18n_provider.instance_variable_set(:@crowdin, crowdin)
+      end
+
+      it 'triggers the creation of the directory on Crowding' do
+        expect(crowdin).to receive(:add_directory).with(crowdin_directory)
+        i18n_provider.add_directory_if_missing
+      end
+    end
+  end
+
+  describe '#request_build' do
+    before do
+      i18n_provider.instance_variable_set(:@crowdin, crowdin)
+    end
+
+    it 'triggers the export of translations' do
+      expect(crowdin).to receive(:export_translations)
+      i18n_provider.request_build
+    end
+  end
+
+  describe '#download_locales' do
```

Comment:
Prefer double-quoted strings unless you need single quotes to avoid extra backslashes for escaping.